Documentation/technical/evolve.txt

```diff
@@ -0,0 +1,885 @@
+If the branch name is omitted, all local changes are listed.
+
+The -r argument list all remote changes.
+
+The -m argument renames a change, given its old and new name.
+
+The -d argument deletes a change. This is one way to resolve divergence.
+
+The -n argument renames the current change, or creates a change of the given
+name for the current commit if no such change exists yet. If given an optional
+commit hash, the change is created for that commit rather than head. If there
+are multiple local changes for the same commit and they are all aliases for the
+same metacommit hash, they are all deleted except the newly-created name. If
+given the name of a metacommit, the new change points to that metacommit.
+
+The --purge argument deletes all obsolete changes and all changes that are
+present in the given branch. Note that such changes can be recovered from the
+reflog.
+
+Combined with the GC protection that is offered, this is intended to facilitate
+a workflow that relies on changes instead of branches. Users could choose to
+work with no local branches and use changes instead - both for mailing list and
+gerrit workflows.
+
+Log
+---
+When a commit is shown in git log that is part of a change, it is decorated with
+extra change information. If it is the head of a change, the name of the change
+is shown next to the list of branches. If it is obsolete, it is decorated with
+the word “obsolete”.
+
+Obslog
+------
+Obslog command lists the change history for the current commit.
+
+Rebase
+------
+In general the rebase command is treated as a modify command. When a change is
+rebased, the new commit replaces the original.
+
+Rebase --abort is special. Its intent is to restore git to the state it had
+prior to running rebase. It should move back any changes to point to the refs
+they had prior to running rebase and delete any new changes that were created as
+part of the rebase. To achieve this, rebase will save the state of all changes
+in refs/metas prior to running rebase and will restore the entire namespace
+after rebase completes (deleting any newly-created changes). Newly-created
+metacommits are left in place, but will have no effect until garbage collected
+since metacommits are only used if they are reachable from refs/metas.
+
+Other options considered
+========================
+We considered several other options for storing the obsolescence graph. This
+section describes the other options and why they were rejected.
+
+Commit header
+-------------
+Add an “obsoletes” field to the commit header that points backwards from a
+commit to the previous commits it obsoletes.
+
+Pros:
+- Very simple
+- Easy to traverse from a commit to the previous commits it obsoletes.
+Cons:
+- Adds a cost to the storage format, even for commits where the change history
+  is uninteresting.
+- Unconditionally prevents the change history from being garbage collected.
+- Always causes the change history to be shared when pushing or pulling changes.
```

Comment:
I don't think this is true either. However, as I said somewhere up above, I'd like to better understand why you want withholding local change history to be a thing when pushing branches.